[5.0.0-alpha.2] Using version string in MySqlServerVersion or MariaDbServerVersion constructor results in NullReferenceException later in MySqlOptions.ApplyDefaultDataTypeMappings()

### Steps to reproduce

I'm attempting to create a new migration via command line.

### The issue

```
Exception message:
Stack trace:

dotnet ef migrations add migration-test-1
Build started...
Build succeeded.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Pomelo.EntityFrameworkCore.MySql.Internal.MySqlOptions.ApplyDefaultDataTypeMappings(MySqlDefaultDataTypeMappings defaultDataTypeMappings, MySqlConnectionSettings connectionSettings)
   at Pomelo.EntityFrameworkCore.MySql.Internal.MySqlOptions.Initialize(IDbContextOptions options)
   at Microsoft.EntityFrameworkCore.Internal.SingletonOptionsInitializer.EnsureInitialized(IServiceProvider serviceProvider, IDbContextOptions options)
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.<>c__DisplayClass4_0.<GetOrAdd>g__BuildServiceProvider|3()
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.<>c__DisplayClass4_0.<GetOrAdd>b__2(Int64 k)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.GetOrAdd(IDbContextOptions options, Boolean providerRequired)
   at Microsoft.EntityFrameworkCore.DbContext..ctor(DbContextOptions options)
   at Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityUserContext`5..ctor(DbContextOptions options)
   at Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityDbContext`8..ctor(DbContextOptions options)
   at Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityDbContext`3..ctor(DbContextOptions options)
   at Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityDbContext`1..ctor(DbContextOptions options)
   at Alugueis.Web.Data.AppDbContext..ctor(DbContextOptions`1 options, IHttpContextAccessor httpContextAccessor, IWebHostEnvironment env) in /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Data/AppDbContext.cs:line 28
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetServiceOrCreateInstance(IServiceProvider provider, Type type)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.<>c__DisplayClass13_3.<FindContextTypes>b__11()
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Object reference not set to an instance of an object.
```
### Further technical details

Pomelo MySQL 5.0.0-alpha.2
DotNet v5.0.100
EF Core v5
EF Command Line tools v5.0
MySQL v8.0.22
O.S. Pop-OS 20.04 (based on Ubuntu 20.04)

Connection code:
```
            services.AddDbContext<AppDbContext>(options =>
                options.UseMySql(
                        connectionString,
                        new MySqlServerVersion("8.0.22"),
                        mySqlOptions => mySqlOptions
                            .CharSetBehavior(CharSetBehavior.NeverAppend))
                    .EnableSensitiveDataLogging()
                    .EnableDetailedErrors()
```

Here is the output with the verbose option
```
Using project '/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj'.
Using startup project '/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj'.
Writing '/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/obj/Alugueis.Web.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=/tmp/tmpT0z2MA.tmp /verbosity:quiet /nologo /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj
Writing '/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/obj/Alugueis.Web.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=/tmp/tmpChSfKy.tmp /verbosity:quiet /nologo /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj
Build started...
dotnet build /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj /verbosity:quiet /nologo
/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Data/AppDbContext.cs(23,33): warning CS0169: O campo "AppDbContext._connectionString" nunca é usado [/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj]
/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Data/AppDbContext.cs(22,47): warning CS0169: O campo "AppDbContext._httpContextAccessor" nunca é usado [/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj]

Compilação com êxito.

/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Data/AppDbContext.cs(23,33): warning CS0169: O campo "AppDbContext._connectionString" nunca é usado [/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj]
/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Data/AppDbContext.cs(22,47): warning CS0169: O campo "AppDbContext._httpContextAccessor" nunca é usado [/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Alugueis.Web.csproj]
    2 Aviso(s)
    0 Erro(s)

Tempo Decorrido 00:01:10.95
Build succeeded.
dotnet exec --depsfile /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/bin/Debug/net5.0/Alugueis.Web.deps.json --additionalprobingpath /home/rafael/.nuget/packages --runtimeconfig /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/bin/Debug/net5.0/Alugueis.Web.runtimeconfig.json /home/rafael/.dotnet/tools/.store/dotnet-ef/5.0.0/dotnet-ef/5.0.0/tools/netcoreapp3.1/any/tools/netcoreapp2.0/any/ef.dll migrations add migration-test-1 --assembly /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/bin/Debug/net5.0/Alugueis.Web.dll --startup-assembly /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/bin/Debug/net5.0/Alugueis.Web.dll --project-dir /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/ --language C# --working-dir /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web --verbose --root-namespace Alugueis.Web
Using assembly 'Alugueis.Web'.
Using startup assembly 'Alugueis.Web'.
Using application base '/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/bin/Debug/net5.0'.
Using working directory '/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web'.
Using root namespace 'Alugueis.Web'.
Using project directory '/media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/'.
Remaining arguments: .
Finding DbContext classes...
Finding IDesignTimeDbContextFactory implementations...
Finding application service provider in assembly 'Alugueis.Web'...
Finding Microsoft.Extensions.Hosting service provider...
Using environment 'Development'.
Using application service provider from Microsoft.Extensions.Hosting.
Found DbContext 'AppDbContext'.
Finding DbContext classes in the project...
System.NullReferenceException: Object reference not set to an instance of an object.
   at Pomelo.EntityFrameworkCore.MySql.Internal.MySqlOptions.ApplyDefaultDataTypeMappings(MySqlDefaultDataTypeMappings defaultDataTypeMappings, MySqlConnectionSettings connectionSettings)
   at Pomelo.EntityFrameworkCore.MySql.Internal.MySqlOptions.Initialize(IDbContextOptions options)
   at Microsoft.EntityFrameworkCore.Internal.SingletonOptionsInitializer.EnsureInitialized(IServiceProvider serviceProvider, IDbContextOptions options)
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.<>c__DisplayClass4_0.<GetOrAdd>g__BuildServiceProvider|3()
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.<>c__DisplayClass4_0.<GetOrAdd>b__2(Int64 k)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.GetOrAdd(IDbContextOptions options, Boolean providerRequired)
   at Microsoft.EntityFrameworkCore.DbContext..ctor(DbContextOptions options)
   at Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityUserContext`5..ctor(DbContextOptions options)
   at Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityDbContext`8..ctor(DbContextOptions options)
   at Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityDbContext`3..ctor(DbContextOptions options)
   at Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityDbContext`1..ctor(DbContextOptions options)
   at Alugueis.Web.Data.AppDbContext..ctor(DbContextOptions`1 options) in /media/rafael/data/dev/csharp/Alugueis/Alugueis.Web/Data/AppDbContext.cs:line 26
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetServiceOrCreateInstance(IServiceProvider provider, Type type)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.<>c__DisplayClass13_3.<FindContextTypes>b__11()
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Object reference not set to an instance of an object.
```

I'll give it a try using local compiled binaries and post results here.
Thanks

--- a/src/EFCore.MySql/Infrastructure/MariaDbServerVersion.cs
+++ b/src/EFCore.MySql/Infrastructure/MariaDbServerVersion.cs
@@ -10,6 +10,7 @@ namespace Microsoft.EntityFrameworkCore
 {
     public class MariaDbServerVersion : ServerVersion
     {
+        public static readonly string MariaDbTypeIdentifier = nameof(ServerType.MariaDb).ToLowerInvariant();
         public static readonly ServerVersion LatestSupportedServerVersion = new MariaDbServerVersion(new Version(10, 5, 5));
 
         public override ServerVersionSupport Supports { get; }
@@ -21,7 +22,7 @@ namespace Microsoft.EntityFrameworkCore
         }
 
         public MariaDbServerVersion(string versionString)
-            : this(FromString(versionString))
+            : this(FromString(versionString, ServerType.MariaDb))
         {
         }
 
@@ -29,10 +30,12 @@ namespace Microsoft.EntityFrameworkCore
             : base(serverVersion.Version, serverVersion.Type, serverVersion.TypeIdentifier)
         {
             if (Type != ServerType.MariaDb ||
-                !string.Equals(TypeIdentifier, nameof(ServerType.MariaDb).ToLowerInvariant()))
+                !string.Equals(TypeIdentifier, MariaDbTypeIdentifier))
             {
                 throw new ArgumentException($"{nameof(MariaDbServerVersion)} is not compatible with the supplied server type.");
             }
+
+            Supports = new MariaDbServerVersionSupport(this);
         }
 
         public class MariaDbServerVersionSupport : ServerVersionSupport

--- a/src/EFCore.MySql/Infrastructure/MySqlServerVersion.cs
+++ b/src/EFCore.MySql/Infrastructure/MySqlServerVersion.cs
@@ -10,6 +10,7 @@ namespace Microsoft.EntityFrameworkCore
 {
     public class MySqlServerVersion : ServerVersion
     {
+        public static readonly string MySqlTypeIdentifier = nameof(ServerType.MySql).ToLowerInvariant();
         public static readonly ServerVersion LatestSupportedServerVersion = new MySqlServerVersion(new Version(8, 0, 21));
 
         public override ServerVersionSupport Supports { get; }
@@ -21,7 +22,7 @@ namespace Microsoft.EntityFrameworkCore
         }
 
         public MySqlServerVersion(string versionString)
-            : this(FromString(versionString))
+            : this(FromString(versionString, ServerType.MySql))
         {
         }
 
@@ -29,10 +30,12 @@ namespace Microsoft.EntityFrameworkCore
             : base(serverVersion.Version, serverVersion.Type, serverVersion.TypeIdentifier)
         {
             if (Type != ServerType.MySql ||
-                !string.Equals(TypeIdentifier, nameof(ServerType.MySql).ToLowerInvariant()))
+                !string.Equals(TypeIdentifier, MySqlTypeIdentifier))
             {
                 throw new ArgumentException($"{nameof(MySqlServerVersion)} is not compatible with the supplied server type.");
             }
+
+            Supports = new MySqlServerVersionSupport(this);
         }
 
         public class MySqlServerVersionSupport : ServerVersionSupport

--- a/src/EFCore.MySql/Infrastructure/ServerVersion.cs
+++ b/src/EFCore.MySql/Infrastructure/ServerVersion.cs
@@ -74,11 +74,11 @@ namespace Microsoft.EntityFrameworkCore
             return FromString(serverVersion);
         }
 
-        public static ServerVersion FromString(string versionString)
+        public static ServerVersion FromString(string versionString, ServerType? serverType = null)
         {
             Check.NotEmpty(versionString, nameof(versionString));
 
-            if (!TryFromString(versionString, out var serverVersion))
+            if (!TryFromString(versionString, serverType, out var serverVersion))
             {
                 throw new InvalidOperationException($"Unable to determine server version from version string '${versionString}'.");
             }
@@ -87,6 +87,9 @@ namespace Microsoft.EntityFrameworkCore
         }
 
         public static bool TryFromString(string versionString, out ServerVersion serverVersion)
+            => TryFromString(versionString, null, out serverVersion);
+
+        public static bool TryFromString(string versionString, ServerType? serverType, out ServerVersion serverVersion)
         {
             Check.NotEmpty(versionString, nameof(versionString));
 
@@ -95,17 +98,15 @@ namespace Microsoft.EntityFrameworkCore
             var semanticVersion = _versionRegex.Matches(versionString);
             if (semanticVersion.Count > 0)
             {
-                var type = versionString.ToLower()
-                    .Contains("mariadb")
-                    ? ServerType.MariaDb
-                    : ServerType.MySql;
-                var version = type == ServerType.MariaDb && semanticVersion.Count > 1
-                    ? Version.Parse(
-                        semanticVersion[1]
-                            .Value)
-                    : Version.Parse(
-                        semanticVersion[0]
-                            .Value);
+                var type = serverType ??
+                           (versionString.ToLower().Contains(MariaDbServerVersion.MariaDbTypeIdentifier)
+                               ? ServerType.MariaDb
+                               : ServerType.MySql);
+
+                var version = type == ServerType.MariaDb &&
+                              semanticVersion.Count > 1
+                    ? Version.Parse(semanticVersion[1].Value)
+                    : Version.Parse(semanticVersion[0].Value);
 
                 serverVersion = type switch
                 {

--- a/test/EFCore.MySql.Tests/MySqlDbContextOptionsExtensionsTest.cs
+++ b/test/EFCore.MySql.Tests/MySqlDbContextOptionsExtensionsTest.cs
@@ -106,5 +106,239 @@ namespace Pomelo.EntityFrameworkCore.MySql
 
             Assert.Equal(MySqlBooleanType.TinyInt1, mySqlOptions.DefaultDataTypeMappings.ClrBoolean);
         }
+
+        [Fact]
+        public void UseMySql_with_MySqlServerVersion_Version()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                new MySqlServerVersion(new Version(8, 0, 21)));
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(8, 0, 21), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MySql, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mysql", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MySqlServerVersion_string_version_only()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                new MySqlServerVersion("8.0.21"));
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(8, 0, 21), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MySql, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mysql", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MySqlServerVersion_string_version_full()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                new MySqlServerVersion("8.0.21-mysql"));
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(8, 0, 21), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MySql, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mysql", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MySqlServerVersion_ServerVersion()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                new MySqlServerVersion(new MySqlServerVersion(new Version(8, 0, 21))));
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(8, 0, 21), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MySql, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mysql", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MySqlServerVersion_incorrect_ServerVersion_throws()
+        {
+            Assert.Throws<ArgumentException>(
+                () =>
+                {
+                    var builder = new DbContextOptionsBuilder();
+
+                    builder.UseMySql(
+                        "Server=foo",
+                        new MySqlServerVersion(new MariaDbServerVersion("10.5.5-mariadb")));
+                });
+        }
+
+        [Fact]
+        public void UseMySql_with_MySqlServerVersion_LatestSupportedServerVersion()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                MySqlServerVersion.LatestSupportedServerVersion);
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(MySqlServerVersion.LatestSupportedServerVersion.Version, mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MySql, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mysql", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MariaDbServerVersion_Version()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                new MariaDbServerVersion(new Version(10, 5, 5)));
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(10, 5, 5), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MariaDb, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mariadb", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MariaDbServerVersion_string_version_only()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                new MariaDbServerVersion("10.5.5"));
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(10, 5, 5), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MariaDb, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mariadb", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MariaDbServerVersion_string_version_full()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                new MariaDbServerVersion("10.5.5-mariadb"));
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(10, 5, 5), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MariaDb, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mariadb", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MariaDbServerVersion_ServerVersion()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                new MariaDbServerVersion(new MariaDbServerVersion(new Version(10, 5, 5))));
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(10, 5, 5), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MariaDb, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mariadb", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_MariaDbServerVersion_incorrect_ServerVersion_throws()
+        {
+            Assert.Throws<ArgumentException>(
+                () =>
+                {
+                    var builder = new DbContextOptionsBuilder();
+
+                    builder.UseMySql(
+                        "Server=foo",
+                        new MariaDbServerVersion(new MySqlServerVersion("8.0.21-mysql")));
+                });
+        }
+
+        [Fact]
+        public void UseMySql_with_MariaDbServerVersion_LatestSupportedServerVersion()
+        {
+            var builder = new DbContextOptionsBuilder();
+
+            builder.UseMySql(
+                "Server=foo",
+                MariaDbServerVersion.LatestSupportedServerVersion);
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(MariaDbServerVersion.LatestSupportedServerVersion.Version, mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MariaDb, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mariadb", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_ServerVersion_FromString()
+        {
+            var builder = new DbContextOptionsBuilder();
+            var serverVersion = ServerVersion.FromString("8.0.21-mysql");
+
+            builder.UseMySql(
+                "Server=foo",
+                serverVersion);
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(new Version(8, 0, 21), mySqlOptions.ServerVersion.Version);
+            Assert.Equal(ServerType.MySql, mySqlOptions.ServerVersion.Type);
+            Assert.Equal("mysql", mySqlOptions.ServerVersion.TypeIdentifier);
+        }
+
+        [Fact]
+        public void UseMySql_with_ServerVersion_AutoDetect()
+        {
+            var builder = new DbContextOptionsBuilder();
+            var serverVersion = ServerVersion.AutoDetect(AppConfig.ConnectionString);
+
+            builder.UseMySql(
+                "Server=foo",
+                serverVersion);
+
+            var mySqlOptions = new MySqlOptions();
+            mySqlOptions.Initialize(builder.Options);
+
+            Assert.Equal(serverVersion.Version, mySqlOptions.ServerVersion.Version);
+            Assert.Equal(serverVersion.Type, mySqlOptions.ServerVersion.Type);
+            Assert.Equal(serverVersion.TypeIdentifier, mySqlOptions.ServerVersion.TypeIdentifier);
+        }
     }
 }

